src/compiler/checker.ts

```diff
@@ -4525,11 +4525,19 @@ namespace ts {
             if (strictNullChecks && declaration.initializer && !(getFalsyFlags(checkExpressionCached(declaration.initializer)) & TypeFlags.Undefined)) {
                 type = getTypeWithFacts(type, TypeFacts.NEUndefined);
             }
-            return declaration.initializer ?
+            return declaration.initializer && !parentDeclarationHasTypeAnnotation(declaration) ?
                 getUnionType([type, checkExpressionCached(declaration.initializer)], UnionReduction.Subtype) :
                 type;
         }
 
+        function parentDeclarationHasTypeAnnotation(binding: BindingElement) {
+            let node: Node = binding;
+            while (node.parent && isBindingPattern(node.parent)) {

```

Comment:
`node.parent` test should be unnecessary.
Following is a bit more strongly typed (no need to test `couldHaveType`):
```ts
let node = binding.parent;
while (isBindingElement(node.parent)) {
    node = node.parent.parent;
}
return !!node.parent.type;
```

But actually, shouldn't we no longer access `.type` and use `getEffectiveTypeAnnotationNode` instead?

Reply:
Good catch on getEffectiveTypeAnnotationNode!

Unfortunately, I tried your proposed change and `node` ends up 1-too-high after the loop -- that is, it's the node above the ParameterDeclaration in the tests below, so `node.parent.type` ends up looking for the type two nodes above the ParameterDeclaration. Also I still had to use `Node` as the type for binding.parent, since `node.parent.parent` is a big union type that isn't assignable to BindingPattern. The elaboration points out that CallSignatureDeclaration, at least, is wrong.

I did work around the 2-too-high problem by tracking the previous parent on each iteration:

```ts
            let node: Node = binding.parent;
            let decl = binding.parent.parent;
            while (isBindingPattern(node)) {
                decl = node.parent;
                node = node.parent.parent;
            }
            return !!getEffectiveTypeAnnotationNode(decl);
```

But this looks worse to me than the original.

Reply:
That can't be right -- if `node.parent.parent` is the node *above* the `ParameterDeclaration`, then `node.parent` must be the `ParameterDeclaration`. But we just tested `isBindingElement(node.parent)`.

Following passed tests when I tried it:
```ts
function parentDeclarationHasTypeAnnotation(binding: BindingElement) {
    let node = binding.parent;
    while (isBindingElement(node.parent)) {
        node = node.parent.parent;
    }
    return !!getEffectiveTypeAnnotationNode(node.parent);
}
```
Node `node` has a static type of `BindingPattern`. The node above the `ParameterDeclaration` would be some kind of `SignatureDeclaration`.